### ISSUE TITLE
Documentation and general usage - Optimizer - differential_evolution

### ISSUE THREAD

**Issue by Mjvolk3** [ICB-DCM/pyPESTO#1025]
When using `pypesto.optimize.ScipyDifferentialEvolutionOptimizer`, I have previously specified initial guess points by adding `x_guesses` to the `pypesto.problem`, but when looking at the options for `scipy.optimize.differential_evolution`, I see that there are options for `x0` and `init`. I am worried about conflicting options. What is the best way to understanding argument priority, clashing options, etc. ? From looking at the API documentation  for `pypesto.optimize.ScipyDifferentialEvolutionOptimizer` it wasn't readily obvious to me. Thanks in advance.

**Comment by PaulJonasJost** [comment]
So if you have defined `x_guesses`in the problem, use `pypesto.optimize.ScipyDifferentialEvolutionOptimizer` and call `pypesto.optimize.minimze`, the following will happen:
* in [pypesto/optimize/optimize.py line 89-93](https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/optimize/optimize.py#L89-L93), if you have not defined a startpointmethod, it will use uniform as start point method, which of the class [`FunctionStartpoints`](https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/startpoint/base.py#L204), a subclass of [`CcheckedStartpoints`](https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/startpoint/base.py#L52).
* In [lines 105-109 of optimize](https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/optimize/optimize.py#L105-L109), th start points are assigned, which uses as a call argument the problem.
* This [call function](https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/startpoint/base.py#L78), given you default options will return `x_guesses`, which will then be supplied to the optimizer as argument of `x_0` [here](https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/optimize/optimize.py#L134-L141)

I think that is the argument line here. I hope this helps in clarifying it :)

**Comment by Mjvolk3** [comment]
This does help, but I have been doing the same thing, walking through code to figure out how potentially conflicting options are handled. For instance, I thought that it would send the `x_guesses` in the `PyswarmsOptimizer` to act as `init_pos` for the particles in the swarm. But instead first `x0` is set by `startpoint`.

https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/optimize/optimize.py#L133-L141

Then `x0` is used only to define `dimension`, not `init_pos`.

https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/optimize/optimizer.py#L886-L891

I think some documentation describing options could be useful, but if it's too much work we can always figure things out by looking at src.

**Comment by PaulJonasJost** [comment]
> So if you have defined `x_guesses`in the problem, use `pypesto.optimize.ScipyDifferentialEvolutionOptimizer` and call `pypesto.optimize.minimze`, the following will happen:
> 
> * in [pypesto/optimize/optimize.py line 89-93](https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/optimize/optimize.py#L89-L93), if you have not defined a startpointmethod, it will use uniform as start point method, which of the class [`FunctionStartpoints`](https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/startpoint/base.py#L204), a subclass of [`CheckedStartpoints`](https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/startpoint/base.py#L52).
> * In [lines 105-109 of optimize](https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/optimize/optimize.py#L105-L109), the start points are assigned, which uses as a call argument the problem.
> * This [call function](https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/startpoint/base.py#L78), given you use default options will return `x_guesses`, which will then be supplied to the optimizer as argument of `x_0` [here](https://github.com/ICB-DCM/pyPESTO/blob/13ce29bfd32a5170284f0d7049ccffc7d1ad7b5d/pypesto/optimize/optimize.py#L134-L141)
> 
> I think that is the argument line here. I hope this helps in clarifying it :)

Though I might have to correct myself, I think it is not passed on, as it is lost in the minimise function. Thanks a lot for the question, will fix that and also clarify in the documentation.

**Comment by PaulJonasJost** [comment]
will also have a look at the PySwarms optimizer 👍🏼

**Comment by Mjvolk3** [comment]
@PaulJonasJost  I've taken some notes on my exploration. They might be useful to you. Excuse them for their brevity. I think the last example showing `generate_swarm` will be informative.

- Say we set `pypesto.problem.x_guesses` ($n \times d$ array). Then the following loop will occur $n$ times, with random starts when using the `PySwarms` optimizer. This works well as we will have particles starting in new locations.

```python
for startpoint, id in zip(startpoints, ids):
        task = OptimizerTask(
            optimizer=optimizer,
            problem=problem,
            x0=startpoint,
            id=id,
            history_options=history_options,
            optimize_options=options,
        )
        tasks.append(task)
```

- If the PySwarms option `init_pos=x_guesses` then there will still be $n$ loops but then the number of particles `par_popsize` will be ignored, and be `len(x_guesses)`, and the initialization will be the same for each new set of particles. This would not be desireable for properly running the optimizer. Because of this is probably best to advise against using `init_pos`.

- **side note**: `x0` is evaluated for both `init_pos=x_guesses` and the default `init_pos=None`.

- I've tested how particles are generated to show this. If we specify the `init_pos`, `n_particles` is overwritten and we only get an initialization of the size of `init_pos`.

```python
>>> from pyswarms.backend.generators import generate_swarm
>>> init_pos = np.ones((2,3))
>>> generate_swarm(n_particles=4, dimensions=init_pos.shape[1],init_pos=init_pos)
array([[1., 1., 1.],
       [1., 1., 1.]])
```

The `dimension` arg needs to be specified, but is overwritten by kwarg `init_pos`.

```python
>>> generate_swarm(n_particles=4, dimensions=5,init_pos=init_pos)
array([[1., 1., 1.],
       [1., 1., 1.]])
```

`init_pos=None` is normal default behavior, with random initialization of particles. Doing this we actually get 4 particles.

```python
>>> generate_swarm(n_particles=4, dimensions=init_pos.shape[1],init_pos=None)
array([[0.17696656, 0.03424968, 0.55395097],
       [0.85859682, 0.5250245 , 0.41581642],
       [0.64526319, 0.04922589, 0.31136249],
       [0.53644966, 0.48004648, 0.32024561]])
```

**Comment by PaulJonasJost** [comment]
Many thanks 🙏🏼  A fix for the `ScipyDifferentialEvolutionOptimizer` as well as at least a warning for the pyswarms optimizer will be released within this week :)

**Comment by PaulJonasJost** [comment]
merged at least warning for optimizers in #1027. Option to pass values to pyswarms will be added in a next commit.
